src/firmware/guest/types/snp.rs

```diff
@@ -117,14 +128,300 @@ bitfield! {
 /// migration agent associated with it, the REPORT_ID_MA is filled in with the report ID of the
 /// migration agent.
 ///
-/// The firmware signs the attestation report with its VCEK. The firmware uses the system wide
+/// The firmware signs the attestation report with its VEK (VCEK or VLEK). The firmware uses the system wide
 /// ReportedTcb value as the TCB version to derive the VCEK. This value is set by the hypervisor.
+/// The VLEK is generated externally and has to be loaded into the machine.
 ///
 /// The firmware guarantees that the ReportedTcb value is never greater than the installed TCB
 /// version
+///
+/// Since the release of the 1.56 ABI, the Attestation Report was bumped from version 2 to 3.
+/// Due to content differences, both versions are kept separately in order to provide backwards compatibility and most reliable security.
+#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
+pub enum AttestationReport {
+    /// Version 2 of the Attestation Report
+    V2(AttestationReportV2),
+    /// Version 3 of the Attestation Report
+    V3(AttestationReportV3),
+}
+
+impl TryFrom<&[u8]> for AttestationReport {
+    type Error = AttestationReportError;
+
+    fn try_from(raw_report: &[u8]) -> Result<Self, Self::Error> {
+        let version =
+            u32::from_le_bytes([raw_report[0], raw_report[1], raw_report[2], raw_report[3]]);
+        // Return the appropriate report version
+        match version {
+            2 => {
+                let report_v2: AttestationReportV2 = raw_report.try_into()?;
+                Ok(AttestationReport::V2(report_v2))
+            }
+            3 => {
+                let report_v3: AttestationReportV3 = raw_report.try_into()?;
+                Ok(AttestationReport::V3(report_v3))
+            }
+            _ => Err(AttestationReportError::UnsupportedReportVersion(version))?,
+        }
+    }
+}
+
+impl Attestable for AttestationReport {
+    fn measurable_bytes(&self) -> io::Result<Vec<u8>> {
+        //Return measurable bytes for the report
+        match self {
+            Self::V2(report) => {
+                let measurable_bytes: &[u8] = &bincode::serialize(report).map_err(|e| {
+                    Error::new(
+                        ErrorKind::Other,
+                        format!("Unable to serialize bytes: {}", e),
+                    )
+                })?;
+                Ok(measurable_bytes[..0x2a0].to_vec())
+            }
+            Self::V3(report) => {
+                let measurable_bytes: &[u8] = &bincode::serialize(report).map_err(|e| {
+                    Error::new(
+                        ErrorKind::Other,
+                        format!("Unable to serialize bytes: {}", e),
+                    )
+                })?;
+                Ok(measurable_bytes[..0x2a0].to_vec())
+            }
+        }
+    }
```

Comment:
Perhaps `Attestable` can be `impl` for the underlying report versions as well?
```suggestion
impl Attestable for AttestationReport {
    fn measurable_bytes(&self) -> io::Result<Vec<u8>> {
        //Return measurable bytes for the report
        match self {
           Self::V2(v2) => v2.measurable_bytes()?
            Self::V3(v3) => v3.measurable_bytes()?
        }
    }
```

Reply:
Good catch!